add Scala functor example

0lejk4 opened the pull request:
Provided Functor example for a Scala language. Functor is implemented using typeclass in Scala and usually is provided by some library like Cats or Scalaz where it is much more complicated than in this simple example.

Commit:
add Scala functor example

0lejk4 commented:
Sorry, didn't see that someone already provided with Scala examples. Should check Pull Requests first next time) But just in case i`ll reopen it.

tshemsedinov commented:
@DzyubSpirit can you please review this?

DzyubSpirit commented:
Yeah, sure :)

DzyubSpirit commented:
This PR tries different from #9. I would say #9 approach is more like Javascript and this PR is more like Haskell in terms of functor definition. #9 and Javascript just implement map method in every functor type (Maybe, Either, Array...). This PR and Haskell typeclasses additionally create Functor entity which is an abstraction of more concrete types (Maybe, Either, Array...)
That makes this PR approach more difficult but at the same time more type safe.
I think both of the #9 should be landed as it is and this code should go into advanced folder

DzyubSpirit commented on Scala/Maybe2.scala:
```diff
@@ -0,0 +1,19 @@
+  import Maybe._
+  import FunctorSyntax._
+
+  print(just("2").fmap(_.toDouble).fmap(a => a + 3))
```
```suggestion
  println(just("2").fmap(_.toDouble).fmap(a => a + 3))
```

DzyubSpirit commented on Scala/Maybe1.scala:
```diff
@@ -0,0 +1,43 @@
+}
+
+//Maybe typeclass that represents optional value. Value type is covariant
+sealed trait Maybe[+T]
```
I, as a Haskeller, like Maybe-Just-Nothing names but as far as I know for Scala Option-Some-None naming is more expected. I think for Scala beginners it would be better to see the same naming and not considering Maybe and Option as two separate concepts

0lejk4 replied:
For sure Option would be better as it is part of std lib. Good point!

DzyubSpirit commented on Scala/Maybe2.scala:
```diff
@@ -0,0 +1,19 @@
+  }
+}
+
+object Maybe2 extends App {
```
```suggestion
object Main2 extends App {
```

DzyubSpirit replied:
Or change Main to Maybe1

DzyubSpirit commented on Scala/Maybe1.scala:
```diff
@@ -0,0 +1,43 @@
+  }
+}
+
+object Main extends App {
```
```suggestion
object Main1 extends App {
```

DzyubSpirit replied:
Or Maybe1

0lejk4 replied:
Yeah, didn`t see this difference, will change it

DzyubSpirit commented on Scala/Maybe1.scala:
```diff
@@ -0,0 +1,43 @@
+package maybe
```
It would really help Scala noobies (me for example) to have comments containing commands to run both examples. For this example it's probably:
// scalac *.scala -d maybe.jar
// scala -cp maybe.jar maybe.Main1

DzyubSpirit replied:
Also it would be great if we can submit both Scala PRs so please create a directory 'advanced' and put your advanced examples in it ;)

0lejk4 replied:
I`ll leave more info in README. About directories: should it be like Scala/Advanced or Advanced/Scala ?

DzyubSpirit replied:
Scala/Advanced

Commit:
Refactor Names and add README

0lejk4 commented:
Just added README, refactored names, moved to Scala/Advanced folder and added comments about the second example, please check it once more. @DzyubSpirit

DzyubSpirit commented on Scala/Advanced/2-functor-syntax.scala:
```diff
@@ -0,0 +1,30 @@
+   * that could be fixed by making Functor typeclass Invariant like it is fixed in Cats
+   * but it is a little bit complicated for our simple example
+   */
+  print(some("2").fmap(_.toDouble).fmap(a => a + 3))
```
```suggestion
  println(some("2").fmap(_.toDouble).fmap(a => a + 3))
```
Except this all looks good. References to the articles in README are espesially great!

0lejk4 replied:
Yeah, forgot about this one, now it`s ready!

Commit:
change print to println

DzyubSpirit commented:
Good work!
